be/src/exprs/map_apply_expr.cpp

```diff
@@ -0,0 +1,141 @@
+// Copyright 2021-present StarRocks, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "exprs/map_apply_expr.h"
+
+#include <fmt/format.h>
+
+#include "column/chunk.h"
+#include "column/column_helper.h"
+#include "column/const_column.h"
+#include "column/fixed_length_column.h"
+#include "column/map_column.h"
+#include "column/vectorized_fwd.h"
+#include "exprs/anyval_util.h"
+#include "exprs/expr_context.h"
+#include "exprs/function_helper.h"
+#include "exprs/lambda_function.h"
+#include "runtime/user_function_cache.h"
+#include "storage/chunk_helper.h"
+
+namespace starrocks {
+MapApplyExpr::MapApplyExpr(const TExprNode& node) : Expr(node, false) {}
+
+MapApplyExpr::MapApplyExpr(TypeDescriptor type) : Expr(std::move(type), false) {}
+
+StatusOr<ColumnPtr> MapApplyExpr::evaluate_checked(ExprContext* context, Chunk* chunk) {
+    std::vector<ColumnPtr> input_columns;
+    NullColumnPtr input_null_map = nullptr;
+    std::shared_ptr<MapColumn> input_map = nullptr;
+    // step 1: get input columns from map(key_col, value_col)
+    for (int i = 1; i < _children.size(); ++i) {
+        ColumnPtr child_col = EVALUATE_NULL_IF_ERROR(context, _children[i], chunk);
+        // the column is a null literal.
+        if (child_col->only_null()) {
+            return ColumnHelper::align_return_type(child_col, type(), chunk->num_rows(), true);
+        }
+        // no optimization for const columns.
+        child_col = ColumnHelper::unpack_and_duplicate_const_column(child_col->size(), child_col);
+        auto data_column = child_col;
+        if (child_col->is_nullable()) {
+            auto nullable = std::dynamic_pointer_cast<NullableColumn>(child_col);
```

Comment:
replace by `down_cast<>`?

Reply:
dynamic_cast may bad performance, use it sparingly if you have an explicit type

Reply:
ok